napari/layers/points/_slice.py

```diff
@@ -0,0 +1,88 @@
+
+
+@dataclass(frozen=True)
+class _PointSliceRequest:
+    """A callable that stores all the input data needed to slice a Points layer.
+
+    This should be treated a deeply immutable structure, even though some
+    fields can be modified in place. It is like a function that has captured
+    all its inputs already.
+
+    In general, the calling an instance of this may take a long time, so you may
+    want to run it off the main thread.
+
+    Attributes
+    ----------
+    dims : _SliceInput
+        Describes the slicing plane or bounding box in the layer's dimensions.
+    data : Any
+        The layer's data field, which is the main input to slicing.
+    dims_indices : Tuple[Union[int, slice], ...]
+        The slice indices in the layer's data space.
+    size : array like
+        Size of each point. This is used in calculating visibility.
+    others
+        See the corresponding attributes in `Layer` and `Image`.
+    """
+
+    dims: _SliceInput
+    data: Any = field(repr=False)
+    dims_indices: Any = field(repr=False)
+    size: Any = field(repr=False)
+    out_of_slice_display: bool = field(repr=False)
+
+    def __call__(self) -> _PointSliceResponse:
+        slice_indices, scale = self._get_slice_data()
+
+        return _PointSliceResponse(indices=slice_indices, scale=scale)
+
+    def _get_slice_data(self):
+        """This method does the actual slicing work."""
+        not_disp = list(self.dims.not_displayed)
+        # We want a numpy array so we can use fancy indexing with the non-displayed
+        # indices, but as self.dims_indices can (and often/always does) contain slice
+        # objects, the array has dtype=object which is then very slow for the
+        # arithmetic below. As Points._round_index is always False, we can safely
+        # convert to float to get a major performance improvement.
+        not_disp_indices = np.array(self.dims_indices)[not_disp].astype(float)
+        if len(self.data) > 0:
+            if self.out_of_slice_display and self.dims.ndim > 2:
+                distances = abs(self.data[:, not_disp] - not_disp_indices)
+                sizes = self.size[:, not_disp] / 2
+                matches = np.all(distances <= sizes, axis=1)
+                size_match = sizes[matches]
+                size_match[size_match == 0] = 1
+                scale_per_dim = (size_match - distances[matches]) / size_match
+                scale_per_dim[size_match == 0] = 1
+                scale = np.prod(scale_per_dim, axis=1)
+                slice_indices = np.where(matches)[0].astype(int)
+                return slice_indices, scale
+            else:
+                data = self.data[:, not_disp]
+                distances = np.abs(data - not_disp_indices)
+                matches = np.all(distances <= 0.5, axis=1)
+                slice_indices = np.where(matches)[0].astype(int)
+                return slice_indices, 1
+        else:
+            return [], np.empty(0)
```

Comment:
Optional: given we're touching this code, I think there's an opportunity for a little unwrapping of this nest loop.

Specifically, when `data` is empty, we could have an early return at the top of function independent of everything else. That allows us to handle the regular and out-of-slice display cases. I also wonder if we should add an early return if `not_disp` is empty (i.e. we're not really slicing) and if that actually matters (or if that optimization is handled elsewhere).

Up to you if you want to pursue this, and how you break this up. If it were me, I'd probably have the conditions directly in the `__call__` function,  and make a couple of helper methods for the actual slicing.